File: tianshu/core/sandbox/docker.py
from __future__ import annotations

import asyncio
import json
import shlex
from pathlib import Path
from typing import Any

from tianshu.config import PROJECT_ROOT
from tianshu.core.log import get_logger

logger = get_logger("sandbox.docker")

IMAGE_NAME = "tianshu-sandbox"
DOCKERFILE = """FROM python:3.12-slim

RUN pip install --no-cache-dir -i https://pypi.tuna.tsinghua.edu.cn/simple \\
    fastapi uvicorn pydantic pydantic-settings httpx typer rich \\
    prompt-toolkit beautifulsoup4 pyyaml aiosqlite pytest

WORKDIR /workspace
CMD ["bash"]
"""


async def _run_docker(args: list[str], timeout: int = 60) -> tuple[str, str | None]:
    try:
        proc = await asyncio.create_subprocess_exec(
            "docker", *args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await asyncio.wait_for(proc.communicate(), timeout=timeout)
    except FileNotFoundError:
        return "", "docker 命令不存在"
    except asyncio.TimeoutError:
        return "", "(docker 操作超时)"
    out = stdout.decode(errors="replace").strip()
    err = stderr.decode(errors="replace").strip()
    if proc.returncode != 0:
        return out, err or f"docker 返回码 {proc.returncode}"
    return out, None


async def docker_available() -> bool:
    out, err = await _run_docker(["info", "--format", "{{.ServerVersion}}"], timeout=15)
    return err is None


async def ensure_image(force_rebuild: bool = False) -> str | None:
    if not force_rebuild:
        out, err = await _run_docker(["image", "inspect", IMAGE_NAME, "--format", "{{.Id}}"], timeout=15)
        if err is None:
            return IMAGE_NAME
    df_path = PROJECT_ROOT / "tianshu" / "core" / "sandbox" / "Dockerfile"
    df_path.write_text(DOCKERFILE, encoding="utf-8")
    out, err = await _run_docker(
        ["build", "-t", IMAGE_NAME, "-f", str(df_path), str(df_path.parent)],
        timeout=300,
    )
    if err is not None:
        logger.error("沙箱镜像构建失败: %s", err)
        return None
    logger.info("沙箱镜像构建成功")
    return IMAGE_NAME


async def run_in_docker(
    args: list[str],
    cwd: Path,
    timeout: int = 30,
    env_extra: dict[str, str] | None = None,
) -> tuple[str, str | None]:
    image = await ensure_image()
    if image is None:
        return "", "Docker 沙箱不可用: 镜像构建失败"
    rel = str(cwd.resolve())
    cmd = ["run", "--rm", "--network", "none", "--memory", "512m", "--cpus", "1"]
    cmd += ["-v", f"{rel}:/workspace:rw", "-w", "/workspace"]
    cmd += ["--name", f"ts-sandbox-{asyncio.current_task().get_name()[:24]}"]
    cmd += [image]
    cmd += args
    out, err = await _run_docker(cmd, timeout=timeout + 15)
    if err is not None:
        return out, f"Docker 执行失败: {err}"
    return out or "(无输出)", None

File: tiankui-ts/src/core/sandbox/docker.ts
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

const execFileP = promisify(execFile);

const HERE = path.dirname(fileURLToPath(import.meta.url));
const IMAGE_NAME = process.env.TIANKUI_SANDBOX_IMAGE ?? "lordofstars/tiankui-sandbox";
const DOCKERFILE = fs.readFileSync(path.join(HERE, "Dockerfile"), "utf-8");

interface DockerRun {
  out: string;
  err: string | null;
}

async function runDocker(args: string[], timeout = 60, killCmd: string[] | null = null): Promise<DockerRun> {
  try {
    const proc = execFileP("docker", args, { timeout: timeout * 1000, maxBuffer: 16 * 1024 * 1024 });
    const { stdout, stderr } = await proc;
    const err = stderr.trim();
    return { out: stdout.trim(), err: err || null };
  } catch (e) {
    const error = e as NodeJS.ErrnoException & { code?: string };
    if (error.code === "ENOENT") return { out: "", err: "docker 命令不存在" };
    const killed = (e as Error & { killed?: boolean }).killed;
    if (killed && killCmd) {
      try {
        await execFileP("docker", killCmd, { timeout: 10000 });
      } catch {
        /* noop */
      }
      return { out: "", err: "(docker 操作超时,已强杀)" };
    }
    if (error.code === "ETIMEDOUT") {
      if (killCmd) {
        try {
          await execFileP("docker", killCmd, { timeout: 10000 });
        } catch {
          /* noop */
        }
      }
      return { out: "", err: "(docker 操作超时,已强杀)" };
    }
    return { out: "", err: (e as Error).message };
  }
}

export async function dockerAvailable(): Promise<boolean> {
  const r = await runDocker(["info", "--format", "{{.ServerVersion}}"], 15);
  return r.err === null;
}

let _imageChecked = false;
let _imageOk = false;

async function ensureImage(forceRebuild = false): Promise<string | null> {
  if (!forceRebuild && _imageChecked) return _imageOk ? IMAGE_NAME : null;
  if (!forceRebuild) {
    const r = await runDocker(["image", "inspect", IMAGE_NAME, "--format", "{{.Id}}"], 15);
    if (r.err === null) {
      _imageChecked = true;
      _imageOk = true;
      return IMAGE_NAME;
    }
  }
  const dfPath = path.join(HERE, "Dockerfile");
  fs.writeFileSync(dfPath, DOCKERFILE, "utf-8");
  const build = await runDocker(["build", "-t", IMAGE_NAME, "-f", dfPath, HERE], 300);
  _imageChecked = true;
  if (build.err !== null) {
    _imageOk = false;
    return null;
  }
  _imageOk = true;
  return IMAGE_NAME;
}

export async function runInDocker(
  args: string[],
  cwd: string,
  timeout = 30,
  envExtra: Record<string, string> | null = null,
): Promise<{ out: string; err: string | null }> {
  const image = await ensureImage();
  if (image === null) return { out: "", err: "Docker 沙箱不可用: 镜像构建失败" };
  const rel = path.resolve(cwd);
  const name = `ts-sandbox-${Date.now().toString(36)}-${Math.random().toString(36).slice(2, 8)}`;
  const cmd = ["run", "--rm", "--network", "none", "--memory", "512m", "--cpus", "1"];
  cmd.push("-v", `${rel}:/workspace:rw`, "-w", "/workspace");
  if (envExtra) {
    for (const [k, v] of Object.entries(envExtra)) cmd.push("-e", `${k}=${v}`);
  }
  cmd.push("--name", name, image, ...args);
  const r = await runDocker(cmd, timeout + 15, ["docker", "kill", name]);
  if (r.err !== null) return { out: r.out, err: `Docker 执行失败: ${r.err}` };
  return { out: r.out || "(无输出)", err: null };
}